torch/utils/data/datapipes/_typing.py

```diff
@@ -368,12 +371,22 @@ def conditional_reset(*args, **kwargs):
             @functools.wraps(setstate_func)
             def wrap_setstate(*args, **kwargs):
                 r"""
-                Set `_restored` to True during `__setstate__`, such that the next `reset()` call during
+                Set `_SnapshotState` to `NotStarted` during `__setstate__`, such that the next `reset()` call during
                 iterator creation will not actually reset the state of the DataPipe.
                 """
-                datapipe = args[0]
-                datapipe._restored = True
-                return setstate_func(*args, **kwargs)
+                try:
+                    return setstate_func(*args, **kwargs)
+                finally:
+                    # This needs happen after because the state being passed in
+                    # may have `datapipe._SnapshotState = Iterating`.
+                    datapipe = args[0]
+                    # FIXME: Without the `if` condition below, the code snippet below would fail. We have
+                    #   investigated but no clue why. Note that this doesn't cause any functional issue.
+                    #   >>> dp = IterableWrapper(range(10))
+                    #   >>> copied_dp = pickle.loads(pickle.dumps(dp))
+                    #   >>> pickle.dumps(dp) == pickle.dumps(copied_dp)
+                    if datapipe._snapshot_state != _SnapshotState.NotStarted:
+                        datapipe._snapshot_state = _SnapshotState.NotStarted
```

Comment:
@VitalyFedyunin you may want to take a closer look at this `setstate` hook. Somehow the bytes from `pickle.dumps` changes when we set `datapipe._snapshot_state = _SnapshotState.NotStarted`, even was `datapipe._snapshot_state` was already `_SnapshotState.NotStarted` prior.

This doesn't cause any functional difference/issue.

@ejguan and I tried to look into why but no luck. Please have a quick look if you can.

Reply:
That's weird. Have you tried to type check datapipe._snapshot_state after restoration. It might become `int` and be serialized differently next time (just guessing).

Reply:
On a second thought, given we decided that restoration is possible even when `datapipe._snapshot_state == _SnapshotState.Iterating`. This `setstate` hook is no longer necessary. All the tests will still pass as it is even without the hook.

The proposed changed here is - do NOT set `datapipe._snapshot_state = _SnapshotState.NotStarted` during `__setstate__`. When a DataPipe graph gets serialized/deserialized, its `_snapshot_state` will not change.

cc: @ejguan